File: backend-go/internal/service/user_service.go
package service

import (
	"context"
	"time"

	"golang.org/x/crypto/bcrypt"

	"github.com/scguoi/remote-ci/backend-go/internal/common"
	"github.com/scguoi/remote-ci/backend-go/internal/dao"
)

// UserService defines the user service interface
// Corresponds to UserService in the Java project
type UserService interface {
	CreateUser(ctx context.Context, req *common.CreateUserRequest, createdBy *string) (*common.UserResponse, error)
	GetUserByID(ctx context.Context, id int64) (*common.UserResponse, error)
	GetUserByUsername(ctx context.Context, username string) (*common.UserResponse, error)
	GetUserByEmail(ctx context.Context, email string) (*common.UserResponse, error)
	UpdateUser(
		ctx context.Context,
		id int64,
		req *common.UpdateUserRequest,
		updatedBy *string,
	) (*common.UserResponse, error)
	DeleteUser(ctx context.Context, id int64, version int) error
	ListUsers(ctx context.Context, filter *common.UserFilter) (*common.PagedResponse[*common.UserResponse], error)
	CheckUsernameExists(ctx context.Context, username string, excludeID *int64) (bool, error)
	CheckEmailExists(ctx context.Context, email string, excludeID *int64) (bool, error)
}

// userServiceImpl implements the UserService interface
// Corresponds to UserServiceImpl in the Java project
type userServiceImpl struct {
	userDAO dao.UserDAO
}

// NewUserService creates a new UserService instance
func NewUserService(userDAO dao.UserDAO) UserService {
	return &userServiceImpl{
		userDAO: userDAO,
	}
}

// CreateUser creates a new user
func (s *userServiceImpl) CreateUser(
	ctx context.Context,
	req *common.CreateUserRequest,
	createdBy *string,
) (*common.UserResponse, error) {
	// Validate username uniqueness
	exists, err := s.userDAO.ExistsByUsername(ctx, req.Username, nil)
	if err != nil {
		return nil, err
	}
	if exists {
		return nil, common.NewUsernameExistsError(req.Username)
	}

	// Validate email uniqueness
	exists, err = s.userDAO.ExistsByEmail(ctx, req.Email, nil)
	if err != nil {
		return nil, err
	}
	if exists {
		return nil, common.NewEmailExistsError(req.Email)
	}

	// Hash password
	passwordHash, err := s.hashPassword(req.Password)
	if err != nil {
		return nil, common.NewInternalError("Failed to hash password", err)
	}

	// Create user entity
	now := time.Now()
	user := &common.User{
		Username:     req.Username,
		Email:        req.Email,
		FullName:     req.FullName,
		PasswordHash: passwordHash,
		PhoneNumber:  req.PhoneNumber,
		IsActive:     true, // Default to active
		CreatedAt:    now,
		UpdatedAt:    now,
		CreatedBy:    createdBy,
		UpdatedBy:    createdBy,
	}

	// Save user
	err = s.userDAO.Create(ctx, user)
	if err != nil {
		return nil, err
	}

	return user.ToResponse(), nil
}

// GetUserByID retrieves a user by ID
func (s *userServiceImpl) GetUserByID(ctx context.Context, id int64) (*common.UserResponse, error) {
	user, err := s.userDAO.GetByID(ctx, id)
	if err != nil {
		return nil, err
	}

	return user.ToResponse(), nil
}

// GetUserByUsername retrieves a user by username
func (s *userServiceImpl) GetUserByUsername(ctx context.Context, username string) (*common.UserResponse, error) {
	user, err := s.userDAO.GetByUsername(ctx, username)
	if err != nil {
		return nil, err
	}

	return user.ToResponse(), nil
}

// GetUserByEmail retrieves a user by email
func (s *userServiceImpl) GetUserByEmail(ctx context.Context, email string) (*common.UserResponse, error) {
	user, err := s.userDAO.GetByEmail(ctx, email)
	if err != nil {
		return nil, err
	}

	return user.ToResponse(), nil
}

// UpdateUser updates an existing user
func (s *userServiceImpl) UpdateUser(
	ctx context.Context,
	id int64,
	req *common.UpdateUserRequest,
	updatedBy *string,
) (*common.UserResponse, error) {
	user, err := s.userDAO.GetByID(ctx, id)
	if err != nil {
		return nil, err
	}

	if user.Version != req.Version {
		return nil, common.NewVersionMismatchError()
	}

	if err := s.applyEmailUpdate(ctx, id, req, user); err != nil {
		return nil, err
	}

	s.applyFieldUpdates(req, user)
	s.applyAudit(updatedBy, user)

	if err := s.userDAO.Update(ctx, user); err != nil {
		return nil, err
	}

	return user.ToResponse(), nil
}

// applyEmailUpdate checks email uniqueness if requested and applies it to user.
func (s *userServiceImpl) applyEmailUpdate(
	ctx context.Context,
	id int64,
	req *common.UpdateUserRequest,
	user *common.User,
) error {
	if req.Email == nil || *req.Email == user.Email {
		return nil
	}
	exists, err := s.userDAO.ExistsByEmail(ctx, *req.Email, &id)
	if err != nil {
		return err
	}
	if exists {
		return common.NewEmailExistsError(*req.Email)
	}
	user.Email = *req.Email
	return nil
}

// applyFieldUpdates copies allowed mutable fields from request to entity.
func (s *userServiceImpl) applyFieldUpdates(req *common.UpdateUserRequest, user *common.User) {
	if req.FullName != nil {
		user.FullName = *req.FullName
	}
	if req.PhoneNumber != nil {
		user.PhoneNumber = req.PhoneNumber
	}
	if req.IsActive != nil {
		user.IsActive = *req.IsActive
	}
}

// applyAudit updates audit fields and timestamps.
func (s *userServiceImpl) applyAudit(updatedBy *string, user *common.User) {
	user.UpdatedAt = time.Now()
	user.UpdatedBy = updatedBy
}

// DeleteUser soft deletes a user
func (s *userServiceImpl) DeleteUser(ctx context.Context, id int64, version int) error {
	return s.userDAO.Delete(ctx, id, version)
}

// ListUsers retrieves a paginated list of users
func (s *userServiceImpl) ListUsers(
	ctx context.Context,
	filter *common.UserFilter,
) (*common.PagedResponse[*common.UserResponse], error) {
	// Set default pagination parameters
	if filter.Size <= 0 {
		filter.Size = 10
	}
	if filter.Size > 100 {
		filter.Size = 100
	}
	if filter.Page < 0 {
		filter.Page = 0
	}

	// Get data
	users, total, err := s.userDAO.List(ctx, filter)
	if err != nil {
		return nil, err
	}

	// Convert to response format
	responses := make([]*common.UserResponse, len(users))
	for i, user := range users {
		responses[i] = user.ToResponse()
	}

	// Calculate pagination info
	totalPages := int((total + int64(filter.Size) - 1) / int64(filter.Size))
	if totalPages == 0 {
		totalPages = 1
	}

	return &common.PagedResponse[*common.UserResponse]{
		Content:       responses,
		Page:          filter.Page,
		Size:          filter.Size,
		TotalElements: total,
		TotalPages:    totalPages,
		First:         filter.Page == 0,
		Last:          filter.Page >= totalPages-1,
	}, nil
}

// CheckUsernameExists checks if a username exists
func (s *userServiceImpl) CheckUsernameExists(ctx context.Context, username string, excludeID *int64) (bool, error) {
	return s.userDAO.ExistsByUsername(ctx, username, excludeID)
}

// CheckEmailExists checks if an email exists
func (s *userServiceImpl) CheckEmailExists(ctx context.Context, email string, excludeID *int64) (bool, error) {
	return s.userDAO.ExistsByEmail(ctx, email, excludeID)
}

// hashPassword encrypts a password
func (s *userServiceImpl) hashPassword(password string) (string, error) {
	bytes, err := bcrypt.GenerateFromPassword([]byte(password), bcrypt.DefaultCost)
	if err != nil {
		return "", err
	}
	return string(bytes), nil
}

// VerifyPassword verifies a password (for login authentication)
func (s *userServiceImpl) VerifyPassword(hashedPassword, password string) error {
	return bcrypt.CompareHashAndPassword([]byte(hashedPassword), []byte(password))
}

File: backend-python/app/core/services/user_service.py
"""User business logic service layer."""

from typing import Optional

from sqlalchemy.orm import Session

from ...db.dao.user_dao import UserDAO
from ..models import User
from ..schemas import UserCreate, UserListResponse, UserResponse, UserUpdate
from ..security import hash_password


class UserService:
    """User business logic service."""

    def __init__(self, db: Session):
        self.db = db
        self.user_dao = UserDAO(db)

    def create_user(
        self, user_create: UserCreate, created_by: str = "system"
    ) -> UserResponse:
        """Create user."""
        # Check if username and email already exist
        if self.user_dao.check_username_exists(user_create.username):
            raise ValueError(f"Username '{user_create.username}' already exists")

        if self.user_dao.check_email_exists(user_create.email):
            raise ValueError(f"Email '{user_create.email}' already exists")

        # Encrypt password
        hashed_password = hash_password(user_create.password)

        # Create user object (temporary password handling)
        user_data = user_create.model_copy()

        # Create user record
        db_user = self.user_dao.create_user(user_data, created_by)

        # Update hashed password (needs to be reset here)
        db_user.hashed_password = hashed_password
        self.db.commit()

        return UserResponse.model_validate(db_user)

    def get_user_by_id(self, user_id: int) -> Optional[UserResponse]:
        """Get user by ID."""
        db_user = self.user_dao.get_user_by_id(user_id)
        if not db_user:
            return None
        return UserResponse.model_validate(db_user)

    def get_user_by_username(self, username: str) -> Optional[UserResponse]:
        """Get user by username."""
        db_user = self.user_dao.get_user_by_username(username)
        if not db_user:
            return None
        return UserResponse.model_validate(db_user)

    def check_username_exists(self, username: str) -> bool:
        """Check if username exists."""
        return self.user_dao.check_username_exists(username)

    def check_email_exists(self, email: str) -> bool:
        """Check if email exists."""
        return self.user_dao.check_email_exists(email)

    def list_users(
        self,
        page: int = 0,
        size: int = 10,
        is_active: Optional[bool] = None,
        username: Optional[str] = None,
        email: Optional[str] = None,
    ) -> UserListResponse:
        """Paginated user list query."""
        if size > 100:  # Limit maximum number per page
            size = 100

        users, total = self.user_dao.list_users(page, size, is_active, username, email)

        user_responses = [UserResponse.model_validate(user) for user in users]

        return UserListResponse(total=total, page=page, size=size, users=user_responses)

    def update_user(
        self, user_id: int, user_update: UserUpdate, updated_by: str = "system"
    ) -> Optional[UserResponse]:
        """Update user."""
        # Check email uniqueness (if email is to be updated)
        if user_update.email:
            existing_user = self.user_dao.get_user_by_email(user_update.email)
            if existing_user and existing_user.id != user_id:
                raise ValueError(
                    f"Email '{user_update.email}' is already used by another user"
                )

        db_user = self.user_dao.update_user(user_id, user_update, updated_by)
        if not db_user:
            return None

        return UserResponse.model_validate(db_user)

    def delete_user(
        self, user_id: int, version: int, deleted_by: str = "system"
    ) -> bool:
        """Delete user."""
        return self.user_dao.delete_user(user_id, version, deleted_by)

    def authenticate_user(self, username: str, password: str) -> Optional[User]:
        """User authentication."""
        from ..security import verify_password

        db_user = self.user_dao.get_user_by_username(username)
        if not db_user or not db_user.is_active:
            return None

        if not verify_password(password, db_user.hashed_password):
            return None

        return db_user
